go/pkg/client/cas.go

```diff
@@ -722,53 +904,97 @@ func (c *Client) DownloadFiles(ctx context.Context, execRoot string, outputs map
 		}
 	}
 
-	eg, eCtx := errgroup.WithContext(ctx)
+	ctx := context.Background()
 	for i, batch := range batches {
-		i, batch := i, batch     // https://golang.org/doc/faq#closures_and_goroutines
-		c.casDownloaders <- true // Reserve an downloader thread.
-		eg.Go(func() error {
-			defer func() { <-c.casDownloaders }()
+		i, batch := i, batch // https://golang.org/doc/faq#closures_and_goroutines
+		go func() (err error) {
+			c.casDownloaders.Acquire(ctx, 1)
+			defer c.casDownloaders.Release(1)
 			if i%logInterval == 0 {
 				log.V(2).Infof("%d batches left to download", len(batches)-i)
 			}
 			if len(batch) > 1 {
 				log.V(3).Infof("Downloading batch of %d files", len(batch))
-				bchMap, err := c.BatchDownloadBlobs(eCtx, batch)
+				bchMap, err := c.BatchDownloadBlobs(ctx, batch)
 				if err != nil {
+					afterDownload(batch, reqs, err)
 					return err
 				}
 				for dg, data := range bchMap {
-					out := outputs[dg]
-					perm := c.RegularMode
-					if out.IsExecutable {
-						perm = c.ExecutableMode
-					}
-					if err := ioutil.WriteFile(filepath.Join(execRoot, out.Path), data, perm); err != nil {
-						return err
+					for _, r := range reqs[dg] {
+						perm := c.RegularMode
+						if r.output.IsExecutable {
+							perm = c.ExecutableMode
+						}
+						r.wait <- ioutil.WriteFile(filepath.Join(r.execRoot, r.output.Path), data, perm)
```

Comment:
Would it be possible to streamline where we respond to requests on the wait channel so its all done the same way? Currently sometimes we write to the channel directly, and in other places we use afterDownload.